[cluster-test] Add experiment to test recovery_time

ankushagarwal opened the pull request:
## Summary

This experiment tests the time it takes for a validator to recover after N accounts have been minted (~N transactions in the system) and the validator has its databases wiped

This relies on prometheus to measure the time between a validator startup and the time it starts committing transactions.

I avoided using `HealthCheckRunner` for this experiment because it seems very complicated to setup because of its dependency on `LogTail`

## Test Plan

Ran on my cluster

```
cluster-test-runner ec2-user:~$ ct --run-experiment recovery_time -- --num-accounts-to-mint 10000
Dec 05 00:52:15.109 INFO Discovered 100 peers in kush4 workspace
Dec 05 00:52:15.409 INFO Log tail thread started in 300 ms
Dec 05 00:52:15.410 INFO Will use kush tag for deployment
Dec 05 00:52:16.055 INFO Starting experiment RecoveryTime
Dec 05 00:52:16.866 INFO Completed minting seed accounts
Dec 05 00:52:35.821 INFO Mint is done
Dec 05 00:52:35.821 INFO Stopping 91fa58e1(10.0.1.129)
Dec 05 00:52:47.440 INFO Deleted directories for 91fa58e1(10.0.1.129)
Dec 05 00:52:47.440 INFO DeleteLibraData 91fa58e1(10.0.1.129)
Dec 05 00:53:07.816 INFO Starting instance 91fa58e1(10.0.1.129)
Dec 05 00:53:08.424 INFO Waiting for instance to be up: 91fa58e1(10.0.1.129)
Dec 05 00:53:13.443 INFO Instance 91fa58e1(10.0.1.129) is up. Waiting for it to start committing.
Dec 05 00:55:28.896 INFO Instance started committing after : 135 seconds
Dec 05 00:55:31.742 INFO Experiment finished, waiting until all affected validators recover
Dec 05 00:55:36.753 INFO Experiment completed
```

andll commented on testsuite/cluster-test/src/experiments/recovery_time.rs:
```diff
@@ -0,0 +1,108 @@
+        result
+    }
+
+    fn run(&mut self, context: &mut Context) -> BoxFuture<Result<Option<String>>> {
```
You won't need all those `clone()` calls below if you annotate references with lifetimes:
```
    fn run<'a>(&'a mut self, context: &'a mut Context) -> BoxFuture<'a, Result<Option<String>>> {
```

andll commented:
I think interesting output of this test could be smth like 
```Recovery rate: xxx txn/sec```

ankushagarwal commented:
Updated based on suggestions.

Also, I switched getting the counters from the instance directly instead of prometheus because prometheus scrapes only 1 data point every 15 seconds and the recovery time can be very flaky because of that : https://github.com/libra/libra/pull/1912/files#diff-7b02d78589d80abf11f09fc872857d03R73

andll commented on testsuite/cluster-test/src/instance.rs:
```diff
@@ -69,6 +70,23 @@ impl Instance {
+                return Ok(response);
+            }
+        }
+        Err(format_err!("Could not find counter: {}", counter))
```
It would be good to include response here to see what was wrong

ankushagarwal replied:
Done

andll commented:
@bors-libra delegate+

bors-libra commented:
:v: @ankushagarwal can now approve this pull request

ankushagarwal commented:
@bors-libra r+

bors-libra commented:
:pushpin: Commit 2ab35e7 has been approved by `ankushagarwal`

<!-- @bors-libra r=ankushagarwal 2ab35e7acc6f6336b725a6b9aa26c012db594498 -->

bors-libra commented:
:hourglass: Testing commit 2ab35e7acc6f6336b725a6b9aa26c012db594498 with merge 357f2726f3a3794a4850e2c87b4fb901d02c6ff8...

bors-libra commented:
:sunny: Test successful - [checks-circle_commit_workflow](https://circleci.com/workflow-run/99bc059a-2312-4817-a1be-504eb08164e8?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link)
Approved by: ankushagarwal
Pushing 357f2726f3a3794a4850e2c87b4fb901d02c6ff8 to master...
